### ISSUE TITLE
Installing jupyterlab 4 prerelease errors about Cargo not being installed

### ISSUE THREAD

**Issue by jasongrout** [jupyterlab/jupyterlab#12436]
## Description

Installing the jupyterlab 4.0 prerelease attempts to install y-py, which errors on installing with the message 

```sh
Collecting y-py<1,>=0.2.3
  Using cached y_py-0.4.3.tar.gz (30 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [6 lines of output]
      
      Cargo, the Rust package manager, is not installed or is not on PATH.
      This package requires Rust and Cargo to compile extensions. Install it through
      the system's package manager or via https://rustup.rs/
      
      Checking for Rust toolchain....
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.
```

Do we now depend on the cargo toolchain to be installed for jupyterlab 4?

## Reproduce

1. Do not have cargo installed in your path (apparently)
2. Create a new conda environment with just python: `mamba create -n testenv python`
3. Install jupyterlab prerelease: `pip install --pre jupyterlab`

## Expected behavior

The prerelease is installed.

## Context

- Operating System and version: macos 12.3.1
- Browser and version: n/a
- JupyterLab version: 4.0.0a23

<details><summary>Troubleshoot Output</summary>
<pre>
Paste the output from running `jupyter troubleshoot` from the command line here.
You may want to sanitize the paths in the output.
</pre>
</details>

<details><summary>Command Line Output</summary>
<pre>
$ pip install --pre jupyterlab
Collecting jupyterlab
  Using cached jupyterlab-4.0.0a23-py3-none-any.whl (8.1 MB)
Collecting notebook-shim>=0.1
  Using cached notebook_shim-0.1.0-py3-none-any.whl (13 kB)
Collecting jupyter-core
  Downloading jupyter_core-4.10.0-py3-none-any.whl (87 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 87.3/87.3 KB 1.0 MB/s eta 0:00:00
Collecting jupyter-server<2,>=1.15.6
  Using cached jupyter_server-1.16.0-py3-none-any.whl (343 kB)
Collecting packaging
  Using cached packaging-21.3-py3-none-any.whl (40 kB)
Collecting ipykernel
  Using cached ipykernel-6.13.0-py3-none-any.whl (131 kB)
Collecting jinja2>=3.0.3
  Using cached Jinja2-3.1.1-py3-none-any.whl (132 kB)
Collecting tornado>=6.1.0
  Using cached tornado-6.1-cp310-cp310-macosx_11_0_arm64.whl
Collecting jupyterlab-server<3,>=2.11.1
  Using cached jupyterlab_server-2.12.0-py3-none-any.whl (53 kB)
Collecting y-py<1,>=0.2.3
  Using cached y_py-0.4.3.tar.gz (30 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [6 lines of output]
      
      Cargo, the Rust package manager, is not installed or is not on PATH.
      This package requires Rust and Cargo to compile extensions. Install it through
      the system's package manager or via https://rustup.rs/
      
      Checking for Rust toolchain....
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

</pre>
</details>

**Comment by agoose77** [comment]
For some reason, the resolver has given you an sdist rather than a built wheel (which are available).

**Comment by jasongrout** [comment]
This was on an M1 mac - perhaps that is the problem, that I'm not finding the binary? Thanks @fcollonval for bringing up this possibility.

**Comment by jasongrout** [comment]
I'm happy to post an m1/arm version of y-py, if it helps.

**Comment by agoose77** [comment]
Are you on 3.10? If so, I think it's just that there isn't a `universal2` wheel for 3.10 for some reason. I'm _not_ an expert in packaging, but as far as I understand it's the universal2 tagged wheel that bundle x86 and ARM binaries, and I can only see wheels for 3.8 and 3.9 on PyPI.

**Comment by agoose77** [comment]
Looking at the GHA log, it looks like the runner doesn't find 3.10: https://github.com/y-crdt/ypy/runs/6040511099?check_suite_focus=true

**Comment by jasongrout** [comment]
Yes, sorry, I should have mentioned I was using python 3.10.

**Comment by agoose77** [comment]
Ah no problem! I wonder why 3.10 isn't covered by the build matrix. I suspect it's not intentional, but I'm not knowledgeable enough to know if it's just something obvious!

**Comment by jasongrout** [comment]
Just to confirm, I installed on python 3.9 and it picked up the universal wheel and installed everything great.



> In any case, maybe we can open an issue on the `y-crdt/y-py` repo given that this is on their side?

Yep, I'll close this issue too.

**Comment by fcollonval** [comment]
@davidbrochart would it be possible to build the y-py conda package for Mac M1? It will become critical for releasing JupyterLab v4 / notebook v7.

**Comment by jasongrout** [comment]
The underlying y-py package is now published for Mac M1, however it is version 0.4.x, and jupyterlab requires 0.3.x, so we still cannot install jupyterlab.

Can we upgrade our dependency to y-py 0.4.x?

**Comment by davidbrochart** [comment]
Sorry @fcollonval, I missed this issue.
@jasongrout I'll open a PR to require `y-py>=0.4.6`.
